src/js/libgeo/Prover.js

```diff
@@ -68,21 +74,108 @@ function incidentPC(p, c) {
-                p.incidences.push(c.name);
-                c.incidences.push(p.name);
-            }
+            return (erg.value.real < 0.0000000000001);
         }
     };
 }
 
 function checkConjectures() {
+    var debug = false;
+    if (debug) console.log("conjectures", conjectures.length);
     if (conjectures.length === 0) return;
-    // TODO: we need some randomized proving here:
-    // move free elements and check conjectures a number of times.
-    // For now assume that all conjectures could be verified.
+
+    //backupGeo
+    stateArrays.prover.set(stateIn);
+    // proverTmp holds temporary good configurations
+    stateArrays.proverTmp.set(stateIn);
+
+    var nummoves = 3;
+
+    // filter free objects which are involved in conjectures
+    var involved;
+
+    var recalcInvolved = function() {
+        involved = [];
+        conjectures.forEach(function(con) {
+            var invs = con.getInvolved();
+            var incis;
+            invs = invs.forEach(function(el) {
+                if (involved.indexOf(el) < 0) {
+                    involved.push(el);
+                    // also add incidences of involved objects
+                    incis = findAllIncis(el, []);
+                    incis.forEach(function(i) {
+                        if (involved.indexOf(i) < 0) involved.push(i);
+                    });
+                }
+            });
+        });
+    };
+
+    // recursively find all incidences to an geo object
+    var findAllIncis = function(el, list) {
+        // get all incidences we don't already know
+        var nincis = el.incidences.filter(function(iels) {
+            return list.indexOf(csgeo.csnames[iels]) < 0;
+        });
+        if (nincis.length === 0) return list;
+        // add new incidences
+        nincis.forEach(function(ii) {
+            list.push(csgeo.csnames[ii]);
+        });
+        // recursive call
+        nincis.forEach(function(nel) {
+            return findAllIncis(csgeo.csnames[nel], list);
+        });
+        return list;
+    };
+
+    recalcInvolved();
+
+    // for jshint move the function definition outside loop 
+    var checkCon = function(con) {
+        return con.holds();
+    };
+
+    var emove, nconject = conjectures.length;
+    for (var kk = 0; kk < nummoves; kk++) {
+        for (var oo = 0; oo < involved.length; oo++) {
+            var el = involved[oo];
+            if (!el.pinned && geoOps[el.type].isMovable) {
+                if (debug) console.log("prover: moving element", el.name);
+                // get random move and move free element
+                emove = geoOps[el.type].getRandomMove(el);
+                movepointscr(el, emove.value, emove.type);
+                // backup configuration if tracing succeeded 
+                if (!tracingFailed) stateArrays.proverTmp.set(stateLastGood);
+                // restore last good configuration if not
+                else stateIn.set(stateArrays.proverTmp);
```

Comment:
I think this should not be neccessary. Let me think aloud for a moment, to see whether I'm right.

You are essentially using `proverTmp` as a synonym for `stateLastGood`, I think. The `!tracingFailed` check is already [in `traceMouseAndScripts`](https://github.com/strobelm/CindyJS/blob/d92c1e166205d9b05faeda2c647ec56372a4257c/src/js/libgeo/Tracing.js#L113), the `stateLastGood.set(stateIn)` is [in `stateContinueFromHere`](https://github.com/strobelm/CindyJS/blob/d92c1e166205d9b05faeda2c647ec56372a4257c/src/js/libgeo/Tracing.js#L54) and the `stateIn.set(stateLastGood)` is [in `traceMouseAndScripts`](https://github.com/strobelm/CindyJS/blob/d92c1e166205d9b05faeda2c647ec56372a4257c/src/js/libgeo/Tracing.js#L101) again.

But you are calling [`movepointscr`](https://github.com/strobelm/CindyJS/blob/d92c1e166205d9b05faeda2c647ec56372a4257c/src/js/libgeo/Tracing.js#L130-L139) which does a number of calls to `stateContinueFromHere` that break this use of state buffers. I think it would be better not to go through `movepointscr` but to call `traceMover` directly, so that all the individual moves combine to form a single grand move.

But there is a problem here: we are not moving a single element along a multi-segment path, but instead we are switching between objects after each segment. Which means the default machinery for tracing a multi-segment path is indeed not appropriate here, and we need some new code to throw away the last move before we switch mover.

But the code we have here doesn't do that right either: it does restore the previous state arrays, but we don't have all our state in those arrays. In particular, the `homog` properties of the elements are not updated. We need to recalc those for a suitably chosen set of elements to actually undo a move.